File: src/c/unsafe_hw.c
// 6️⃣ Unsafe Hardware Register Access
// 
// Compile: gcc -O2 -S src/c/unsafe_hw.c -o target/unsafe_hw.s
// Assembly file: target/unsafe_hw.s
//
// Key assembly differences to observe:
// - Direct memory access instructions (mov)
// - volatile keyword prevents compiler optimizations
// - No safety boundaries, entire program can use raw pointers
// - Generated assembly nearly identical to Rust version
// - C has no type system level safety guarantees
// - No distinction between safe and unsafe operations
// - Compiler cannot optimize away volatile operations
// - Potential for accidental unsafe operations anywhere in code

#include <stdio.h>

int main() {
    volatile unsigned int *ptr = (unsigned int *)0x1000;
    *ptr = 42;          // Volatile write prevents compiler optimization
    unsigned int val = *ptr;  // Volatile read prevents compiler optimization
    printf("%u\n", val);
    return 0;
} 

File: src/rust/unsafe_hw.rs
// 6️⃣ Unsafe Hardware Register Access
// 
// Compile: cargo rustc --bin unsafe_hw --release -- --emit=asm
// Assembly file: target/release/deps/unsafe_hw-*.s
//
// Key assembly differences to observe:
// - Direct memory access instructions (mov)
// - Volatile operations prevent compiler optimizations
// - unsafe block isolates dangerous operations
// - Generated assembly nearly identical to C version
// - Rust type system ensures only unsafe blocks can use raw pointers
// - Safety boundaries clearly marked in source code
// - Compiler cannot optimize away volatile operations
// - Zero-cost abstraction: unsafe has no runtime overhead

fn main() {
    let ptr = 0x1000 as *mut u32;
    unsafe {
        // Volatile write prevents compiler optimization
        core::ptr::write_volatile(ptr, 42);
        // Volatile read prevents compiler optimization
        let val = core::ptr::read_volatile(ptr);
        println!("{}", val);
    }
} 